IDE/iotsafe/memory-tls.c

```diff
@@ -0,0 +1,385 @@
+/* memory-tls.c
+ *
+ * Copyright (C) 2006-2020 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL. (formerly known as CyaSSL)
+ *
+ * wolfSSL is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/* IoT-safe client side demo
+    - server uses software crypto and buffers
+    - client uses IoT-Safe
+
+    Client and server communicates in a cooperative
+    scheduling mechanism within the same thread.
+    Two buffers in memory are used for client<=>server communication.
+*/
+
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/ssl.h>
+#include <wolfssl/error-ssl.h>
+#include <wolfssl/wolfcrypt/port/iotsafe/iotsafe.h>
+
+#include <stdio.h>
+#include <string.h>
+
+#include "ca-cert.c"
+
+/* IoTSAFE Certificate slots */
+
+/* File Slot '03' is pre-provisioned with
+ * local certificate.
+ */
+#define CRT_CLIENT_FILE_ID 0x03     /* pre-provisioned */
+
+/* File Slot '04' is pre-provisioned with the
+ * server's EC public key certificate
+ */
+#define CRT_SERVER_FILE_ID 0x04
+
+/* IoTSAFE Key slots */
+
+/* Key slot '02' is pre-provisioned with
+ * the client private key.
+ */
+#define PRIVKEY_ID      0x02 /* pre-provisioned */
+
+/* Key slot '03' is used by wolfSSL to generate
+ * the ECDH key that will be used during the TLS
+ * session.
+ */
+#define ECDH_KEYPAIR_ID 0x03
+
+/* Key slot '04' is used to store the public key
+ * received from the peer.
+ */
+#define PEER_PUBKEY_ID  0x04
+
+/* Key slot '05' is used to store a public key
+ * used for ecc verification
+ */
+#define PEER_CERT_ID    0x05
+
+/* The following define
+ * activates mutual authentication */
+#define CLIENT_AUTH
+
+
+#define CLIENT_IOTSAFE
+#define CA_ECC
+
+/* client messages to server in memory */
+#define TLS_BUFFERS_SZ (1024 * 8)
```

Comment:
Technically TLS supports up to 16KB packets unless using MAX_FRAGMENT extension to communicate smaller sizes. Although it depends on what the peers actually send. Perhaps consider an overridable macro?

Reply:
Agree, but this is just a simple handshake + hello IoTsafe world example, which does not exchange large buffers. ServerSend/ClientSend will return `SSL_ERROR_WANT_WRITE` if the buffer is full.